server/src/test/java/org/elasticsearch/common/util/SingleObjectCacheTests.java

```diff
@@ -0,0 +1,64 @@
+/*
+ * Licensed to Crate.io GmbH ("Crate") under one or more contributor
+ * license agreements.  See the NOTICE file distributed with this work for
+ * additional information regarding copyright ownership.  Crate licenses
+ * this file to you under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.  You may
+ * obtain a copy of the License at
+ *
+ *   http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS, WITHOUT
+ * WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.  See the
+ * License for the specific language governing permissions and limitations
+ * under the License.
+ *
+ * However, if you have executed another commercial license agreement
+ * with Crate these terms will supersede the license and you may use the
+ * software solely pursuant to the terms of the relevant commercial agreement.
+ */
+
+package org.elasticsearch.common.util;
+
+import static org.assertj.core.api.Assertions.assertThat;
+
+import java.util.concurrent.atomic.AtomicInteger;
+
+import org.elasticsearch.test.ESTestCase;
+import org.junit.Test;
+
+import io.crate.common.unit.TimeValue;
+
+public class SingleObjectCacheTests extends ESTestCase {
+
+    @Test
+    public void test_cache_refreshes_after_interval() throws Exception {
+        AtomicInteger numRefreshes = new AtomicInteger();
+        var cache = new SingleObjectCache<>(TimeValue.timeValueNanos(10), 0) {
+
+            @Override
+            protected Integer refresh() {
+                return numRefreshes.incrementAndGet();
+            }
+        };
+        assertBusy(() -> assertThat(cache.getOrRefresh()).isGreaterThan(1));
+    }
+
+    @Test
+    public void test_cache_caches_until_refresh_interval_passed() throws Exception {
+        AtomicInteger numRefreshes = new AtomicInteger();
+        var cache = new SingleObjectCache<>(TimeValue.timeValueHours(1), 0) {
+
+            @Override
+            protected Integer refresh() {
+                return numRefreshes.incrementAndGet();
+            }
+        };
+
+        assertThat(cache.getOrRefresh()).isEqualTo(1);
```

Comment:
This is actually a bit strange that it always refreshes once. It's due to `lastRefreshNanos` not getting initialized. I didn't want to change that in this PR, as this is just about the wall-clock.

Reply:
Pushed a fixup to change it. Otherwise the test can be flaky, depending on the `nanoTime` value.

Reply:
Looks like I was too trigger happy with the fixup. Some components rely on the refresh happening once before the interval is over.

Pushed another fixup that reverts the semantic change and handles a different first value in the test case instead.